### PR TITLE
[fx]add split module pass and unit test from pipeline passes

### PR DESCRIPTION
**Pull request by YuliangLiu0306**

### CODE REVIEW

**Comment by YuliangLiu0306** [comment]
image

File: colossalai/fx/passes/adding_split_node_pass.py [hunk 1]
```diff
@@ -26,8 +26,14 @@ def balanced_split_pass(gm: torch.fx.GraphModule, pp_size: int):
         if accumulate_param_amount >= params_per_partition:
             accumulate_param_amount = 0
             pp_size -= 1
-            with mod_graph.inserting_after(node):
-                split_node = mod_graph.create_node('call_function', pipe_split)
+            # If the next node is output node, we will insert split annotation before
+            # node to make sure there is at least one node in last partitio.
```
Comment (FrankLeeeee) [inline colossalai/fx/passes/adding_split_node_pass.py:30]: partitio -> partition

File: colossalai/fx/passes/split_module.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    Example:
+
+        This is a sample setup:
+
+            import torch
+            from torch.fx.symbolic_trace import symbolic_trace
+            from torch.fx.graph_module import GraphModule
+            from torch.fx.node import Node
+            from torch.fx.passes.split_module import split_module
```
Comment (FrankLeeeee) [inline colossalai/fx/passes/split_module.py:66]: This is still using torch API

File: tests/test_fx/test_pipeline/test_hf_model/hf_utils.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    # set rng state and compute output of split model
+    torch.set_rng_state(cpu_rng_state)
+    output_part0 = model_part0(**kwargs)
+    sig = inspect.signature(model_part1.forward)
+    if isinstance(output_part0, torch.Tensor):
+        output_part1 = model_part1(output_part0)
+    else:
+        if len(output_part0) > len(sig.parameters):
+            output_part0 = output_part0[:len(sig.parameters)]
+        output_part1 = model_part1(*output_part0)
```
Comment (FrankLeeeee) [inline tests/test_fx/test_pipeline/test_hf_model/hf_utils.py:53]: HF output may contain multiple values, will `output_part1` contain the same number of values?

File: tests/test_fx/test_pipeline/test_timm_model/test_timm.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import torch
+import pytest
+try:
+    import timm.models as tm
+except:
+    pass
+from timm_utils import split_model_and_compare_output
+
+
+def test_timm_models_without_control_flow():
```
Comment (FrankLeeeee) [inline tests/test_fx/test_pipeline/test_timm_model/test_timm.py:10]: Skip this as well as the current CI does not install timm

File: tests/test_fx/test_pipeline/test_torchvision/test_torchvision.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import inspect
+
+MANUAL_SEED = 0
+random.seed(MANUAL_SEED)
+np.random.seed(MANUAL_SEED)
+torch.manual_seed(MANUAL_SEED)
+torch.backends.cudnn.deterministic = True
+
+
+def test_torchvision_models():
```
Comment (FrankLeeeee) [inline tests/test_fx/test_pipeline/test_torchvision/test_torchvision.py:21]: Skip this, too.

File: tests/test_fx/test_pipeline/test_timm_model/timm_utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+MANUAL_SEED = 0
+random.seed(MANUAL_SEED)
+np.random.seed(MANUAL_SEED)
+torch.manual_seed(MANUAL_SEED)
+torch.backends.cudnn.deterministic = True
+
+
+def split_model_and_compare_output(model, data, meta_args=None):
+    model.eval()
```
Comment (FrankLeeeee) [inline tests/test_fx/test_pipeline/test_timm_model/timm_utils.py:18]: I am not sure why timm tests will fail for control-flow-based models. You can try to debug this by [adding hooks to modules](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html) to check which modules produce inconsistent outputs.

File: tests/test_fx/test_pipeline/test_hf_model/hf_utils.py (hunk 3 above)
Comment (YuliangLiu0306) [reply tests/test_fx/test_pipeline/test_hf_model/hf_utils.py:53]: Yes, it has same number of elements and numerically equal to HF output, but the output_part1 is type of Tuple[torch.Tensor], not HF output class.

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**